tracing/tracing.go

```diff
@@ -0,0 +1,91 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package tracing
+
+import (
+	gocontext "context"
+
+	"github.com/containerd/containerd/log"
+	srvconfig "github.com/containerd/containerd/services/server/config"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/semconv"
+)
+
+// InitOpenTelemetry does the initialization of open telemetry at startup and sets ups
+// various components as configured in config file if supported.
+func InitOpenTelemetry(ctx gocontext.Context, config *srvconfig.Config) error {
+	if config.OpenTelemetry == *new(srvconfig.OpenTelemetryConfig) {
+		logrus.Info("OpenTelemetry configuration is not found in config, tracing will be disabled")
+		return nil
+	}
+
+	err := config.OpenTelemetry.Validate()
+	if err != nil {
+		return errors.Wrap(err, "invalid open telemetry configuration")
+	}
+
+	exporter, err := InitOpenTelemetryExporter(ctx, config)
+
+	if err != nil || exporter == nil {
+		return errors.Wrapf(err, "failed to initialize exporter")
+	}
+
+	defer func() {
+		err := exporter.Shutdown(ctx)
+		if err != nil {
+			errors.Errorf("failed to stop exporter: %v", err)
+		}
+	}()
```

Comment:
Is it okay to shutdown the exporter at the end of this function?

Reply:
No, this function should probably return this closure instead of deferring it so that it can be called by the main package when the application is ready to shutdown.